pkg/auth/dependency/forgotpassword/provider.go

```diff
@@ -0,0 +1,280 @@
+package forgotpassword
+
+import (
+	"net/url"
+	"path"
+	"time"
+
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/audit"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/hook"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/principal/password"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/urlprefix"
+	"github.com/skygeario/skygear-server/pkg/auth/dependency/userprofile"
+	"github.com/skygeario/skygear-server/pkg/auth/event"
+	"github.com/skygeario/skygear-server/pkg/auth/model"
+	taskspec "github.com/skygeario/skygear-server/pkg/auth/task/spec"
+	"github.com/skygeario/skygear-server/pkg/core/async"
+	"github.com/skygeario/skygear-server/pkg/core/auth/authinfo"
+	"github.com/skygeario/skygear-server/pkg/core/auth/metadata"
+	"github.com/skygeario/skygear-server/pkg/core/config"
+	"github.com/skygeario/skygear-server/pkg/core/mail"
+	"github.com/skygeario/skygear-server/pkg/core/sms"
+	"github.com/skygeario/skygear-server/pkg/core/template"
+	coretime "github.com/skygeario/skygear-server/pkg/core/time"
+)
+
+type Provider struct {
+	AppName                     string
+	EmailMessageConfiguration   config.EmailMessageConfiguration
+	SMSMessageConfiguration     config.SMSMessageConfiguration
+	ForgotPasswordConfiguration *config.ForgotPasswordConfiguration
+
+	Store Store
+
+	AuthInfoStore        authinfo.Store
+	UserProfileStore     userprofile.Store
+	PasswordAuthProvider password.Provider
+	PasswordChecker      *audit.PasswordChecker
+	HookProvider         hook.Provider
+	TimeProvider         coretime.Provider
+	URLPrefixProvider    urlprefix.Provider
+	TemplateEngine       *template.Engine
+	MailSender           mail.Sender
+	SMSClient            sms.Client
+	TaskQueue            async.Queue
+}
+
+// SendCode checks if loginID is an existing login ID.
+// If not found, ErrLoginIDNotFound is returned.
+// If the login ID is not of type email or phone, ErrUnsupportedLoginIDType is returned.
+// Otherwise, a code is generated.
+// The code expires after a specific time.
+// The code becomes invalid if it is consumed.
+// Finally the code is sent to the login ID asynchronously.
+func (p *Provider) SendCode(loginID string) (err error) {
+	// TODO(forgotpassword): Test SendCode
+	// Send single email
+	// Send multiple email
+	// Send email with the normalized email
+	prins, err := p.PasswordAuthProvider.GetPrincipalsByLoginID("", loginID)
+	if err != nil {
+		return
+	}
+
+	if len(prins) <= 0 {
+		err = ErrLoginIDNotFound
+		return
+	}
+
+	for _, prin := range prins {
```

Comment:
After removing realm, is there any case that we need to handle multiple principals in forgot password?

Reply:
Maybe `username` is configured to allow `@` and a username looks like an email address.
But the loop body does not handle this case correct. It should skip non-email and non-phone login ID.